gtsam/slam/PlanarProjectionFactor.h

```diff
@@ -0,0 +1,154 @@
+/**
+ * Similar to GenericProjectionFactor, but:
+ *
+ * * 2d pose variable (robot on the floor)
+ * * constant landmarks
+ * * batched input
+ * * numeric differentiation
+ *
+ * This factor is useful for high-school robotics competitions,
+ * which run robots on the floor, with use fixed maps and fiducial
+ * markers.  The camera offset and calibration are fixed, perhaps
+ * found using PlanarSFMFactor.
+ *
+ * The python version of this factor uses batches, to save on calls
+ * across the C++/python boundary, but here the only extra cost
+ * is instantiating the camera, so there's no batch.
+ *
+ * @see https://www.firstinspires.org/
+ * @see PlanarSFMFactor.h
+ *
+ * @file PlanarSFMFactor.h
+ * @brief for planar smoothing
+ * @date Dec 2, 2024
+ * @author joel@truher.org
+ */
+#pragma once
+
+#include <gtsam/base/Testable.h>
+#include <gtsam/base/Lie.h>
+
+#include <gtsam/geometry/Cal3DS2.h>
+#include <gtsam/geometry/PinholeCamera.h>
+#include <gtsam/geometry/Point3.h>
+#include <gtsam/geometry/Pose2.h>
+#include <gtsam/geometry/Pose3.h>
+#include <gtsam/geometry/Rot3.h>
+#include <gtsam/nonlinear/NonlinearFactor.h>
+#include <gtsam/base/numericalDerivative.h>
+
+
+namespace gtsam {
+    /**
+     * @class PlanarProjectionFactor
+     * @brief Camera projection for robot on the floor.
+     */
+    class PlanarProjectionFactor : public NoiseModelFactorN<Pose2> {
+        typedef NoiseModelFactorN<Pose2> Base;
+        static const Pose3 CAM_COORD;
+
+    protected:
+
+        Point3 landmark_; // landmark
```

Comment:
Before I do a more thorough review, quick sanity check: things that are stored in the factory are typically constant parameters, but here you store a bunch of things. Could you clarify in the PR comment what this factor optimizes for, and what the measurement function is? Be very clear on what is a constant and what is a variable.

Reply:
Also, optimizing over calibration parameters is brave - and makes it rather different from other basic SFM factors

Reply:
OK, I updated the top comment.  You're right that there are a bunch of constants for the projection factor, but really only one more than ProjectionFactor itself: the constant landmark.

Reply:
> Also, optimizing over calibration parameters is brave - and makes it rather different from other basic SFM factors

re: brave, oh, ha ha, it is?  :-)  it does seem to work in simulation.  I haven't tried it on a real robot yet.

re: difference, I think maybe this factor should have a different name anyway, since there's definitely no "S" being extracted from "M," given the constant landmarks.  What would a better name be?

Reply:
> > Also, optimizing over calibration parameters is brave - and makes it rather different from other basic SFM factors
> 
> re: brave, oh, ha ha, it is? :-) it does seem to work in simulation. I haven't tried it on a real robot yet.

It can definitely work. To avoid convergence issues I’d put at least a prior on your calibration parameters.

> re: difference, I think maybe this factor should have a different name anyway, since there's definitely no "S" being extracted from "M," given the constant landmarks. What would a better name be?

You’re right. We sometimes have more than one related factor in one file. I can see several versions here, it should almost be the same effort

- PlanarProjectionFactor1: 1 unknown, the pose
- PlanarProjectionFactor2: conventional: 2 unknowns, the pose and the landmarks. You could typedef this to PlanarSfmFactor.
- PlanarProjectionFactor3: 3 unknowns, including calibration

These can all share a common measurement function: h(pose, point, cal, H1, H2, H3), defined in a PlatProjectionFactorBase

I’d put all of these in one file.